File: typescript/test/conformance.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import {
  ageAt,
  buildQuery,
  computeInvoice,
  ComputedRegistry,
  computeTax,
  ConverterRegistry,
  eraOf,
  evaluateCondition,
  fiscalHalf,
  fiscalQuarter,
  fiscalYear,
  FormatterRegistry,
  isAllowed,
  isBusinessDay,
  nextBusinessDay,
  optionValueIsStale,
  prevBusinessDay,
  tenure,
  visibleOptions,
  type FieldDefinition,
  type FilterDefinition,
  type OptionsOwner,
  ValidatorRegistry,
} from "../src/index.js";

// Shared conformance fixtures (spec/conformance), consumed identically by the
// Dart and Java editions.
const DIR = "../spec/conformance";
const load = (file: string): any[] =>
  JSON.parse(readFileSync(`${DIR}/${file}`, "utf8"));

describe("conformance: formatters", () => {
  const fmt = new FormatterRegistry();
  for (const c of load("formatters.json")) {
    it(`${c.name} ${JSON.stringify(c.value)} ${JSON.stringify(c.options ?? {})}`, () => {
      expect(fmt.format(c.name, c.value, c.options ?? {})).toBe(c.expected);
    });
  }
});

describe("conformance: converters", () => {
  const conv = new ConverterRegistry();
  for (const c of load("converters.json")) {
    it(`${c.name} ${JSON.stringify(c.value)}`, () => {
      expect(String(conv.convert(c.name, c.value))).toBe(String(c.expected));
    });
  }
});

describe("conformance: validators", () => {
  const validators = new ValidatorRegistry();
  for (const c of load("validators.json")) {
    it(`${c.type} ${JSON.stringify(c.value)}`, () => {
      const result = validators.run(c.value, { type: c.type, params: c.params ?? {} });
      expect(result === null).toBe(c.valid);
      if (c.message != null) expect(result).toBe(c.message);
    });
  }
});

describe("conformance: queries", () => {
  for (const c of load("queries.json")) {
    it(JSON.stringify(c.params), () => {
      const search = {
        columns: 1,
        filters: c.filters.map((f: any) => ({
          field: f.field,
          label: f.field,
          type: f.type,
          operator: f.operator,
          options: [],
          config: {},
        })),
      };
      const q = buildQuery(search, c.params);
      const e = c.expected;
      expect(q.conditions.length).toBe(e.conditions.length);
      e.conditions.forEach((ec: any, i: number) => {
        expect(q.conditions[i].field).toBe(ec.field);
        expect(q.conditions[i].operator).toBe(ec.operator);
        expect(String(q.conditions[i].value)).toBe(String(ec.value));
      });
      expect(q.sortField ?? null).toBe(e.sortField);
      expect(q.sortAscending).toBe(e.sortAscending);
      expect(q.page).toBe(e.page);
      expect(q.pageSize).toBe(e.pageSize);
    });
  }
});

describe("conformance: tax", () => {
  for (const c of load("tax.json")) {
    it(`${c.amount}@${c.rate} ${c.rounding ?? "floor"}${c.included ? " inc" : ""}`, () => {
      const r = computeTax(c.amount, {
        rate: c.rate,
        included: c.included === true,
        rounding: c.rounding ?? "floor",
      });
      expect(r).toEqual({ net: c.expected.net, tax: c.expected.tax, gross: c.expected.gross });
    });
  }
});

describe("conformance: fiscal", () => {
  for (const c of load("fiscal.json")) {
    it(`${c.date} sm=${c.startMonth ?? 4}`, () => {
      const sm = c.startMonth ?? 4;
      expect(fiscalYear(c.date, sm)).toBe(c.expected.year);
      expect(fiscalQuarter(c.date, sm)).toBe(c.expected.quarter);
      expect(fiscalHalf(c.date, sm)).toBe(c.expected.half);
    });
  }
});

describe("conformance: age/tenure", () => {
  for (const c of load("age.json")) {
    it(`${c.from} -> ${c.to}`, () => {
      expect(tenure(c.from, c.to)).toEqual({ years: c.years, months: c.months });
      expect(ageAt(c.from, c.to)).toBe(c.years);
    });
  }
});

describe("conformance: era", () => {
  for (const c of load("era.json")) {
    it(`${c.date}`, () => {
      expect(eraOf(c.date)).toEqual({
        name: c.expected.name,
        abbr: c.expected.abbr,
        year: c.expected.year,
      });
    });
  }
});

describe("conformance: invoice", () => {
  for (const c of load("invoice.json")) {
    it(`${c.lines.length} lines${c.included ? " inc" : ""}`, () => {
      const inv = computeInvoice(c.lines, {
        included: c.included === true,
        rounding: c.rounding ?? "floor",
      });
      expect(inv.byRate.length).toBe(c.expected.byRate.length);
      c.expected.byRate.forEach((er: any, i: number) => {
        expect(String(inv.byRate[i].rate)).toBe(String(er.rate));
        expect(inv.byRate[i].net).toBe(er.net);
        expect(inv.byRate[i].tax).toBe(er.tax);
        expect(inv.byRate[i].gross).toBe(er.gross);
      });
      expect(inv.total).toEqual({
        net: c.expected.total.net,
        tax: c.expected.total.tax,
        gross: c.expected.total.gross,
      });
    });
  }
});

describe("conformance: conditions", () => {
  for (const c of load("conditions.json")) {
    it(`${JSON.stringify(c.condition)} on ${JSON.stringify(c.record)} (mode: ${c.mode})`, () => {
      expect(evaluateCondition(c.condition, c.record, c.mode)).toBe(c.expected);
    });
  }
});

describe("conformance: option filter", () => {
  // 選択肢の連動。Dart 版と同じ fixture を食う（定義だけで決まる純粋なロジック）。
  const fixture = JSON.parse(
    readFileSync(`${DIR}/option_filter.json`, "utf8"),
  ) as { cases: any[] };
  for (const c of fixture.cases) {
    it(c.name, () => {
      // `as: filter` のケースは検索条件の形で確認する（判定は共有＝OptionsOwner）。
      const field: OptionsOwner =
        c.as === "filter"
          ? ({
              field: c.field.field,
              label: c.field.field,
              type: "select",
              operator: "equals",
              options: c.field.options,
              optionsFrom: c.field.optionsFrom,
              config: {},
            } satisfies FilterDefinition)
          : ({
              field: c.field.field,
              label: c.field.field,
              type: "select",
              required: false,
              readOnly: false,
              validators: [],
              options: c.field.options,
              optionsFrom: c.field.optionsFrom,
              normalize: [],
              config: {},
              roles: [],
              columns: [],
              rowFields: [],
            } satisfies FieldDefinition);
      expect(visibleOptions(field, c.record).map((o) => o.value)).toEqual(
        c.visible,
      );
      expect(optionValueIsStale(field, c.record)).toBe(c.stale);
    });
  }
});

describe("conformance: computed", () => {
  const reg = new ComputedRegistry();
  for (const c of load("computed.json")) {
    it(`${JSON.stringify(c.computed)} on ${JSON.stringify(c.record)}`, () => {
      expect(reg.compute(c.computed, c.record)).toBe(c.expected);
    });
  }
});

describe("conformance: access", () => {
  for (const c of load("access.json")) {
    it(`${JSON.stringify(c.roles)} / ${JSON.stringify(c.userRoles)}`, () => {
      expect(isAllowed(c.roles, c.userRoles)).toBe(c.expected);
    });
  }
});

describe("conformance: business day", () => {
  for (const c of load("businessday.json")) {
    it(`${c.date} h=${c.holidays.length}`, () => {
      expect(isBusinessDay(c.date, c.holidays)).toBe(c.expected.isBusinessDay);
      expect(nextBusinessDay(c.date, c.holidays)).toBe(c.expected.next);
      expect(prevBusinessDay(c.date, c.holidays)).toBe(c.expected.prev);
    });
  }
});

File: flutter/packages/hatake_core/test/conformance_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:hatake_core/hatake_core.dart';
import 'package:test/test.dart';

/// Runs the shared conformance fixtures (spec/conformance) against the Dart
/// implementation. The same fixtures are consumed by the TS and Java editions,
/// guaranteeing identical output across languages.
const _dir = '../../../spec/conformance';

List<dynamic> _load(String file) =>
    jsonDecode(File('$_dir/$file').readAsStringSync()) as List<dynamic>;

Map<String, dynamic> _loadMap(String file) =>
    jsonDecode(File('$_dir/$file').readAsStringSync()) as Map<String, dynamic>;

Map<String, Object?> _opts(Map<String, dynamic> c) =>
    (c['options'] as Map?)?.cast<String, Object?>() ?? const {};

List<Map<String, Object?>> _rows(Object? raw) => [
      for (final row in raw as List) (row as Map).cast<String, Object?>(),
    ];

/// Numbers are compared as normalized strings so `200` and `200.0` match across
/// languages.
String _num(Object? value) {
  if (value == null) return 'null';
  final n = value as num;
  return n == n.roundToDouble() ? n.toInt().toString() : n.toString();
}

void main() {
  group('conformance: formatters', () {
    final fmt = FormatterRegistry();
    for (final raw in _load('formatters.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['name']} ${c['value']} ${c['options'] ?? ''}', () {
        expect(fmt.format(c['name'] as String, c['value'], _opts(c)),
            c['expected']);
      });
    }
  });

  group('conformance: converters', () {
    final conv = ConverterRegistry();
    for (final raw in _load('converters.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['name']} ${c['value']}', () {
        final result = conv.convert(c['name'] as String, c['value']);
        expect(result.toString(), c['expected'].toString());
      });
    }
  });

  group('conformance: validators', () {
    final validators = ValidatorRegistry();
    for (final raw in _load('validators.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['type']} ${c['value']}', () {
        final def = ValidatorDefinition(
          type: c['type'] as String,
          params: (c['params'] as Map?)?.cast<String, Object?>() ?? const {},
        );
        final result = validators.run(c['value'], def);
        expect(result == null, c['valid']);
        if (c['message'] != null) expect(result, c['message']);
      });
    }
  });

  group('conformance: tax', () {
    for (final raw in _load('tax.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['amount']}@${c['rate']} ${c['rounding'] ?? 'floor'}'
          '${c['included'] == true ? ' inc' : ''}', () {
        final r = computeTax(
          c['amount'] as num,
          rate: c['rate'] as num,
          included: c['included'] == true,
          rounding: c['rounding'] as String? ?? 'floor',
        );
        final e = c['expected'] as Map<String, dynamic>;
        expect(r.net, e['net']);
        expect(r.tax, e['tax']);
        expect(r.gross, e['gross']);
      });
    }
  });

  group('conformance: fiscal', () {
    for (final raw in _load('fiscal.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['date']} sm=${c['startMonth'] ?? 4}', () {
        final sm = (c['startMonth'] as int?) ?? 4;
        final e = c['expected'] as Map<String, dynamic>;
        expect(fiscalYear(c['date'] as String, startMonth: sm), e['year']);
        expect(fiscalQuarter(c['date'] as String, startMonth: sm), e['quarter']);
        expect(fiscalHalf(c['date'] as String, startMonth: sm), e['half']);
      });
    }
  });

  group('conformance: age/tenure', () {
    for (final raw in _load('age.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['from']} -> ${c['to']}', () {
        final t = tenure(c['from'] as String, c['to'] as String);
        expect(t.years, c['years']);
        expect(t.months, c['months']);
        expect(ageAt(c['from'] as String, c['to'] as String), c['years']);
      });
    }
  });

  group('conformance: era', () {
    for (final raw in _load('era.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['date']}', () {
        final ed = eraOf(c['date'] as String);
        final e = c['expected'] as Map<String, dynamic>;
        expect(ed?.name, e['name']);
        expect(ed?.abbr, e['abbr']);
        expect(ed?.year, e['year']);
      });
    }
  });

  group('conformance: invoice', () {
    for (final raw in _load('invoice.json')) {
      final c = raw as Map<String, dynamic>;
      test('${(c['lines'] as List).length} lines'
          '${c['included'] == true ? ' inc' : ''}', () {
        final lines = [
          for (final l in c['lines'] as List)
            InvoiceLine(
              amount: (l as Map)['amount'] as num,
              rate: l['rate'] as num,
            ),
        ];
        final inv = computeInvoice(
          lines,
          included: c['included'] == true,
          rounding: c['rounding'] as String? ?? 'floor',
        );
        final e = c['expected'] as Map<String, dynamic>;
        final eByRate = e['byRate'] as List;
        expect(inv.byRate.length, eByRate.length);
        for (var i = 0; i < eByRate.length; i++) {
          final er = eByRate[i] as Map<String, dynamic>;
          expect(inv.byRate[i].rate.toString(), er['rate'].toString());
          expect(inv.byRate[i].net, er['net']);
          expect(inv.byRate[i].tax, er['tax']);
          expect(inv.byRate[i].gross, er['gross']);
        }
        final et = e['total'] as Map<String, dynamic>;
        expect(inv.total.net, et['net']);
        expect(inv.total.tax, et['tax']);
        expect(inv.total.gross, et['gross']);
      });
    }
  });

  group('conformance: conditions', () {
    for (final raw in _load('conditions.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['condition']} on ${c['record']}', () {
        final cond = (c['condition'] as Map).cast<String, Object?>();
        final record = (c['record'] as Map).cast<String, Object?>();
        expect(evaluateCondition(cond, record), c['expected']);
      });
    }
  });

  group('conformance: computed', () {
    final reg = ComputedRegistry();
    for (final raw in _load('computed.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['computed']} on ${c['record']}', () {
        final computed = (c['computed'] as Map).cast<String, Object?>();
        final record = (c['record'] as Map).cast<String, Object?>();
        final result = reg.compute(computed, record);
        final expected = c['expected'];
        if (expected is num) {
          expect((result as num).toDouble(), expected.toDouble());
        } else {
          expect(result, expected);
        }
      });
    }
  });

  group('conformance: access', () {
    for (final raw in _load('access.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['roles']} / ${c['userRoles']}', () {
        final roles = [for (final r in c['roles'] as List) r.toString()];
        final userRoles = {for (final r in c['userRoles'] as List) r.toString()};
        expect(isAllowed(roles, userRoles), c['expected']);
      });
    }
  });

  group('conformance: dashboard aggregate', () {
    final registry = AggregateRegistry();
    final fixture = _loadMap('dashboard_aggregate.json');
    for (final raw in fixture['aggregate'] as List) {
      final c = raw as Map<String, dynamic>;
      test(c['name'], () {
        final result = registry.aggregate(
          c['op'] as String,
          _rows(c['rows']),
          field: c['field'] as String?,
        );
        expect(_num(result), _num(c['expected']));
      });
    }

    for (final raw in fixture['groupBy'] as List) {
      final c = raw as Map<String, dynamic>;
      test('groupBy: ${c['name']}', () {
        final actual = registry.aggregateBy(
          c['op'] as String,
          _rows(c['rows']),
          labelField: c['labelField'] as String,
          valueField: c['valueField'] as String?,
        );
        expect(
          [for (final b in actual) '${b.label}=${_num(b.value)}'],
          [
            for (final e in c['expected'] as List)
              '${(e as Map)['label']}=${_num(e['value'])}',
          ],
        );
      });
    }
  });

  group('conformance: business day', () {
    for (final raw in _load('businessday.json')) {
      final c = raw as Map<String, dynamic>;
      test('${c['date']} h=${(c['holidays'] as List).length}', () {
        final holidays = {for (final h in c['holidays'] as List) h.toString()};
        final e = c['expected'] as Map<String, dynamic>;
        expect(isBusinessDay(c['date'] as String, holidays: holidays),
            e['isBusinessDay']);
        expect(nextBusinessDay(c['date'] as String, holidays: holidays),
            e['next']);
        expect(prevBusinessDay(c['date'] as String, holidays: holidays),
            e['prev']);
      });
    }
  });
}
